File: 2626-count-the-number-of-good-subarrays/solution.java
class Solution {
    public long countGood(int[] nums, int k) {
        int n = nums.length;
        int i=0;
        int j=0;
        int pairsCount = 0;
        long count = 0;
        Map<Integer, Integer> map = new HashMap<>();

        while(j < n) {
            pairsCount += map.getOrDefault(nums[j], 0);
            map.put(nums[j], map.getOrDefault(nums[j],0)+1);

            while(pairsCount >= k) {
                // Count valid rights
                // count all the elements after first i, at the ending j
                count += n-j;
                map.put(nums[i], map.getOrDefault(nums[i],0)-1);
                pairsCount -= map.getOrDefault(nums[i], 0);
                i++;
            }
            j++;
        }
        return count;
    }
}

File: 2626-count-the-number-of-good-subarrays/solution.swift
class Solution {
    func countGood(_ nums: [Int], _ k: Int) -> Int {
        var dict = [Int: Int]()
        var count = 0
        var n = nums.count
        var i = 0
        var j = 0
        var pairs = 0

        while j < n {
            pairs += dict[nums[j], default: 0]
            dict[nums[j], default: 0] += 1

            while pairs >= k {
                count += n - j
                dict[nums[i], default: 0] -= 1
                pairs -= dict[nums[i], default: 0]
                i += 1
            }
            j += 1
        }
        return count
    }
}
